### PR TITLE
[FLINK-29605] Allow implementers to have bounded and unbounded SourceTestSuiteBase#testSourceMetrics

### PR DESCRIPTION
**Pull request by echauchot**
## What is the purpose of the change

Allow implementers to have bounded and unbounded SourceTestSuiteBase#testSourceMetrics
Based on [timeout pr](https://github.com/apache/flink/pull/21035).

## Brief change log

make SourceTestSuiteBase#testSourceMetrics compatible with both bounded and unbounded sources and add a boundedness parameter to the test to allow the implemeter to pass the corresponding boundedness by overriding this test in the ITCase of its source. Default (without override) is unbounded mode for conservative reasons.

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes test feature
  - If yes, how is the feature documented? javadoc

### CODE REVIEW

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java [hunk 1]
```diff
@@ -462,11 +477,18 @@ public void testSourceMetrics(
                             // skip failed assert try
                             return false;
                         }
-                    });
+                    },
+                    METRICS_TIMEOUT,
+                    DELAY_BETWEEN_CHECKS,
+                    String.format(
+                            "Timeout while comparing source metrics with %d expected value",
+                            getTestDataSize(testRecordCollections)));
```
Comment (zentol) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java:485]: We're trying to get rid of timeouts in tests.

Comment (echauchot) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java:485]: And by the way, this change you're commenting belongs to [the timeout pr](https://github.com/apache/flink/pull/21035) on which this PR is based. We should discuss it there.

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
> ## CI report:
> * [4505e0c](https://github.com/apache/flink/commit/4505e0c5d6384a7faadb84d98b7339bf9bbbb527) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=41947)
> 
> Bot commands

Strange thing. I have set a timeout of 30s and it is hit from time to time by pulsar E2E tests. This means that in some cases metrics retrieval in `testSourceMetrics` is either too long (more than 30s) or the counter of reported source metrics is incorrect leading to infinite wait on the counter condition.

**Comment by echauchot** [comment]
I think `SourceTestSuiteBase#testSourceMetrics` is defined as a streaming test. Closing this PR

**Pull request closed without merging**